src/solver/mha/mha_solver.cpp

```diff
@@ -0,0 +1,306 @@
+
+                void* fp32_ws = getBuffPart(params.GetWorkspace(), 0);
+                void* fp8_ws  = getBuffPart(params.GetWorkspace(), 1);
+                void* prng_ws = getBuffPart(params.GetWorkspace(), 2);
+
+#if MIOPEN_USE_GEMM
+                CallGemmStridedBatched(handle_,
+                                       QK_desc,
+                                       params.GetData().qData,
+                                       0,
+                                       params.GetData().kData,
+                                       0,
+                                       fp32_ws,
+                                       0,
+                                       GemmBackend_t::rocblas);
+#endif
+                decltype(auto) softmax_kernel = handle_.Run(kernels.front());
+                softmax_kernel(fp32_ws,
+                               fp8_ws,
+                               params.GetData().mData,
+                               params.GetData().zInvData,
+                               params.GetData().amaxSData,
+                               params.GetData().descaleQData,
+                               params.GetData().descaleKData,
+                               params.GetData().scaleSData,
+                               prng_ws,
+                               params.GetData().dropoutProbabilityData,
+                               seq_len,
+                               nhs);
+
+#if MIOPEN_USE_GEMM
+                CallGemmStridedBatched(handle_,
+                                       SV_desc,
+                                       fp8_ws,
+                                       0,
+                                       params.GetData().vData,
+                                       0,
+                                       fp32_ws,
+                                       0,
+                                       GemmBackend_t::rocblas);
+#endif
+
+                decltype(auto) scale_reduce_kernel = handle_.Run(kernels.back());
+                scale_reduce_kernel(fp32_ws,
+                                    params.GetData().oData,
+                                    params.GetData().amaxOData,
+                                    params.GetData().descaleSData,
+                                    params.GetData().descaleVData,
+                                    params.GetData().scaleOData,
+                                    nhsd);
+
+                if(profiling)
+                {
+                    hipEventRecord(stop.get(), handle_.GetStream());
+                    handle_.EnableProfiling(true);
+                    hipEventSynchronize(stop.get());
+                    float mS = 0;
+                    hipEventElapsedTime(&mS, start.get(), stop.get());
+                    handle_.ResetKernelTime();
+                    handle_.AccumKernelTime(mS);
+                }
+            };
+        };
+
+    return result;
+}
+
+bool Mha::MayNeedWorkspace() const { return true; }
```

Comment:
not sure if I saw it but `GetSolution()` should also store a size value in `result.workspace_size`.

Reply:
@CAHEK7 @amberhassaan ...except that if the solution does not require workspace, then it may rely on the default (which is 0).